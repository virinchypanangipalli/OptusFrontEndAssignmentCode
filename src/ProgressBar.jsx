import React, { useEffect, useState } from "react";
import axios from "axios";
import Progress from "./Progress";

export const ProgressBar = () => {
  const [bars, setbars] = useState([]);
  const [buttons, setbuttons] = useState([]);
  const [selectedBar, setselectedBar] = useState(0);
  useEffect(() => {
    getbars();
  }, []);

  const handleChange = (e) => {
    setselectedBar(e.target.value);
  };

  const getbars = () => {
    axios
      .get("https://pb-api.herokuapp.com/bars")
      .then((res) => {
        let bars = [];
        res.data.bars.forEach((bar) => {
          let barObj = { bar, background: "lightblue", color: "black" };
          if (typeof bar === "number") {
            bars.push(barObj);
          }
        });

        setbars(bars);
        setbuttons(res.data.buttons);
      })
      .catch((error) => {
        alert(error);
      });
  };
  const handleBtnClick = (val) => {
    let newBar = { ...bars[selectedBar] };
    newBar.bar = newBar.bar + val;
    if (newBar.bar >= 100) {
      newBar.background = "red";
      newBar.color = "white";
    } else {
      if (newBar.bar <= 0) {
        newBar.bar = 0;
      }
      newBar.background = "lightblue";
      newBar.color = "black";
    }
    let allBars = [...bars];
    allBars[selectedBar] = newBar;
    setbars(allBars);
  };

  return (
    <div className="container">
      <h3 className="text-center">Progress Bars Demo</h3>
      <div className="row bars">
        <div className="col-sm"></div>
        <div className="col-sm">
          {bars.length > 0 &&
            bars.map((bar, i) => (
              <div key={i} className="col">
                {typeof bar.bar === "number" && <Progress bar={bar} />}
                {typeof bar.bar !== "number" && (
                  <div>
                    Error Loading Progressbar .. Please Refresh and try again
                  </div>
                )}
              </div>
            ))}
          {bars.length === 0 && <div>Loading....</div>}
          <div className="row buttons">
            <ul className="">
              <li>
                {bars.length > 0 && (
                  <select
                    name="bar"
                    value={selectedBar}
                    onChange={(e) => {
                      handleChange(e);
                    }}
                  >
                    {bars.map((bar, i) => (
                      <option key={i} value={i}>
                        Progress # {i + 1}
                      </option>
                    ))}
                  </select>
                )}{" "}
              </li>
              {buttons.map((button, i) => (
                <li key={i}>
                  <button
                    onClick={() => {
                      handleBtnClick(button, i);
                    }}
                  >
                    {button}
                  </button>
                </li>
              ))}
            </ul>
          </div>
        </div>
        <div className="col"></div>
      </div>
    </div>
  );
};
