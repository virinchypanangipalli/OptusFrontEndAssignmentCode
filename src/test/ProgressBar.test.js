import ProgressBar from "../ProgressBar";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

jest.mock("./api.js", () => {
  return {
    fetchData: jest.fn(),
  };
});

describe("Testing Progress Bar Component", () => {
  let wrapper;
  const getbars = jest.fn();

  beforeAll(() => {
    wrapper = shallow(<ProgressBar getbars={getbars} />);
    expect(wrapper.getElements()).toMatchSnapshot();
    jest.clearAllMocks();
  });

  test("Checking Heading Text", () => {
    expect(wrapper.find("h3").text()).toContain("Progress Bars Demo");
  });

  test("Checking Bars Length ", () => {
    expect(wrapper.find(".bar").length).toEqual(0);
  });

  test("Checking Buttons Length ", () => {
    expect(wrapper.find(".button").length).toEqual(0);
  });
  test("Checking Option Length ", () => {
    expect(wrapper.find(".option").length).toEqual(0);
  });
  it("fetches async data", () => {
    const promise = new Promise((resolve, reject) =>
      setTimeout(
        () =>
          resolve({
            buttons: [16, 22, -13, -24],
            bars: [15, 11, 78, 23],
            limit: 160,
          }),
        100
      )
    );
    const wrapper = shallow(<ProgressBar />);
    promise.then(() => {
      wrapper.update();

      expect(wrapper.find(".bar").length).toEqual(4);
      expect(wrapper.find(".button").length).toEqual(4);
    });
  });
});
