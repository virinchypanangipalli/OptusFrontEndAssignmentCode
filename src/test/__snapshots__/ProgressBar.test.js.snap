// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
Array [
  <div
    className="container"
  >
    <h3
      className="text-center"
      data-testid="heading"
    >
      Progress Bars Demo
    </h3>
    <div
      className="row bars"
    >
      <div
        className="col-sm"
      />
      <div
        className="col-sm"
      >
        <div>
          Loading....
        </div>
        <div
          className="row buttons"
        >
          <ul>
            <li />
          </ul>
        </div>
      </div>
      <div
        className="col"
      />
    </div>
  </div>,
]
`;
